OpenAlex Integration

AdamF42 opened the pull request:
**Please note that this development has been funded by the Vietsch foundation, co-funded by 4Science and the project coordination is provided by University of Cambridge.**

---

## References
* Related to DSpace/dspace-angular#4017

## Description
This PR introduces support for [OpenAlex ](https://docs.openalex.org/how-to-use-the-api/api-overview) as a metadata source for live import and publication claim in DSpace. The integration includes new service configurations, metadata mappings, and improvements to the publication loader to support OpenAlex as a suggestion provider.

## Instructions for Reviewers
### Import new Metadata Fields

Two new metadata fields are added: `dc.identifier.openalex` and `dspace.openalex.lastimport`.  These need to be added to your Metadata Registry.  For example, via Docker:
```
# Reload the "dspace.*" fields
docker container exec -it dspace bash -c "JAVA_TOOL_OPTIONS= && ./bin/dspace registry-loader -metadata config/registries/dspace-types.xml"

# Reload the "dc.*" fields
docker container exec -it dspace bash -c "JAVA_TOOL_OPTIONS= && ./bin/dspace registry-loader -metadata config/registries/dublin-core-types.xml"
```


### How to Test Live Import  

1. **Login** to DSpace and navigate to **MyDSpace**.  
2. Select one of the supported entities (e.g., **Publication**).  
![image(3)](https://github.com/user-attachments/assets/0c584016-3730-4fd1-ba75-94efe831ff77)
3. Choose the **OpenAlex import** option.  (Use the **publication title** unless you have an **OpenAlex author ID**).  
![image(4)](https://github.com/user-attachments/assets/f6f1304c-1e4f-468a-ba6a-4516f43e594a)
4. Select one of the results and click **"Start submission"**.  
![image(5)](https://github.com/user-attachments/assets/0a6c279f-3f3e-4fc5-9e7d-f3d8f6aa00fc)
5. Choose the **collection** where the item should be imported.  
6. Verify that:  
   - The **submission page** opens correctly.  
   - The metadata displayed in the **preview** matches the selected result.
![image(6)](https://github.com/user-attachments/assets/77f7a146-fbe7-4776-99f1-4745d53e4631)

### How to Test [Publication Claim](https://wiki.lyrasis.org/display/DSDOC8x/Publication+Claim)
1. Set Up Required Metadata**  
    Run the following command to add the new **dc-types**:  
    ```bash
    dspace registry-loader -metadata config/registries/dspace-types.xml

2. Import an Author from OpenAlex

    Perform a live import of an author from OpenAlex using the steps above.

3. Create the Researcher Profile

    Create a Researcher Profile by associating the imported author with a newly created EPerson.
    
![image(7)](https://github.com/user-attachments/assets/0fa94eb5-490d-415c-bcf0-6e459b38d024)

4. Run the Import Suggestion Script using openalex as loader
![image(8)](https://github.com/user-attachments/assets/806de5da-aa2e-486e-b409-bd36802322dd)

5. Verify Suggestion Pop-up

    Login as the researcher.
    Confirm that the suggestions pop-up appears in MyDSpace.
![image(9)](https://github.com/user-attachments/assets/62e3db38-a16c-4c46-9b35-76a7cb069a79)

6. Click on **review the suggestions** the get the results
![image(10)](https://github.com/user-attachments/assets/470c1a50-49b4-47df-bce2-32416729adf5)

7. Test Approve & Import

    Click "Approve & Import" on a suggested publication.
    Verify that a workspace item is created successfully in the selected collection.

**Admin View: Review All Suggestions**

Using an admin account, you can see all suggestions from all sources under the section **Notifications > Publication Claim**.

From there, you can select the available suggestion source by clicking on **Actions**.

![admin_pub_claim](https://github.com/user-attachments/assets/6e9ce7a9-f5ef-4f9c-83fd-6db434721259)


Example of OpenAlex Source.
![admin_openalex_pub_claim](https://github.com/user-attachments/assets/6c0cab08-3501-4ee2-a1f6-1da2a74c0578)


## List of changes in this PR:
1. OpenAlex API Integration
    - Added configuration properties to define OpenAlex API endpoints:  
      ```properties
      openalex.url.works = https://api.openalex.org/works
      openalex.url.authors = https://api.openalex.org/authors
      openalex.url.sources = https://api.openalex.org/sources
      openalex.url.institutions = https://api.openalex.org/institutions
      openalex.url.publishers = https://api.openalex.org/publishers
      openalex.url.funders = https://api.openalex.org/funders

    - Implemented OpenAlexImportMetadataSourceServiceImpl, which interacts with OpenAlex to fetch metadata for works, authors, sources, institutions, publishers, funders.
    - Introduced field mapping classes (OpenAlexPublicationFieldMapping and OpenAlexPersonFieldMapping) to convert OpenAlex API responses into ImportRecord objects. Currently mapped entities are:
        - Works → Publications
        - Authors → Persons
        - Sources → Journal (iif type='journal')
        - Institutions → OrgUnit
        - Publishers → OrgUnit
        - Funders → OrgUnit
   
2. Live Import Support
    - Registered OpenAlex as a LiveImportDataProvider, allowing users to fetch metadata in real time from OpenAlex.

3. Refactoring
    - Moved the existing PublicationLoader (previously under the OpenAIRE project) to a new package: suggestion.loader.
        This change makes it reusable for multiple metadata sources.
    - Extended PublicationLoader to create OpenAlexPublicationLoader, which retrieves metadata from OpenAlex.

4. CLI Support for Publication Import

    - Introduced a generic script configuration: PublicationLoaderCliScriptConfiguration, registered as import-loader-suggestions.
    - Implemented PublicationLoaderRunnable as a reusable script runner.
    - The script can now be executed dynamically by specifying a loader argument (e.g., openalex in suggestions.xml).
    - Added a new DC metadata field to store the timestamp of the last import from OpenAlex.



## Checklist

- [x] My PR is **created against the `main` branch** of code (unless it is a backport or is fixing an issue specific to an older branch).
- [ ] My PR is **small in size** (e.g. less than 1,000 lines of code, not including comments & integration tests). Exceptions may be made if previously agreed upon.
- [x] My PR **passes Checkstyle** validation based on the [Code Style Guide](https://wiki.lyrasis.org/display/DSPACE/Code+Style+Guide).
- [x] My PR **includes Javadoc** for _all new (or modified) public methods and classes_. It also includes Javadoc for large or complex private methods.
- [x] My PR **passes all tests and includes new/updated Unit or Integration Tests** based on the [Code Testing Guide](https://wiki.lyrasis.org/display/DSPACE/Code+Testing+Guide).
- [x] My PR **includes details on how to test it**. I've provided clear instructions to reviewers on how to successfully test this fix or feature.
- [x] If my PR includes new libraries/dependencies (in any `pom.xml`), I've made sure their licenses align with the [DSpace BSD License](https://github.com/DSpace/DSpace/blob/main/LICENSE) based on the [Licensing of Contributions](https://wiki.lyrasis.org/display/DSPACE/Code+Contribution+Guidelines#CodeContributionGuidelines-LicensingofContributions) documentation.
- [x] If my PR modifies REST API endpoints, I've opened a separate [REST Contract](https://github.com/DSpace/RestContract/blob/main/README.md) PR related to this change.
- [x] If my PR includes new configurations, I've provided basic technical documentation in the PR itself.
- [x] If my PR fixes an issue ticket, I've [linked them together](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue).

AdamF42 committed:
[CST-18015] Added base OpenAlex Publication integration.

[CST-18015] Added base OpenAlex Publication integration.

AdamF42 committed:
[CST-18015] Added OpenAlex Person

[CST-18015] InvertedIndexProcessor doc and tests

AdamF42 committed:
[CST-18015] Added mapConverter

AdamF42 committed:
[CST-18016] Added OpenAlexPublicationLoader (openaire PublicationLoader TBR)

AdamF42 committed:
[CST-18016] Avoid setting empty dspaceUrl if dspaceSSRUrl is blank

AdamF42 committed:
[CST-18016] Updated SolrSuggestionProvider

AdamF42 committed:
[CST-18016] Fix to PublicationLoader to extends SolrSuggestionProvider

AdamF42 committed:
[CST-18016] Moved PublicationLoader from opeanire package

AdamF42 committed:
[CST-18016] Added OpenAlex external-providers.cfg

AdamF42 committed:
Revert "[CST-18016] Avoid setting empty dspaceUrl if dspaceSSRUrl is blank"

This reverts commit e387573bfc602645a4c9870a97657624ea4b3696.

AdamF42 committed:
[CST-18016] Avoid setting empty dspaceUrl if dspaceSSRUrl is blank

AdamF42 committed:
[CST-18016] Filtered publications by author on OpenAlexImportMetadataSourceServiceImpl

AdamF42 committed:
[CST-18016] Added java doc for OpenAlexImportMetadataSourceServiceImpl

AdamF42 committed:
[CST-18016] Added base tests for OpenAlexImportMetadataSourceServiceImpl

AdamF42 committed:
[CST-18016] Added another OpenAlexImportMetadataSourceServiceImpl to be used in Publication Claim

AdamF42 committed:
[CST-18016] Added JavaDoc to OpenAlexPublicationLoader

AdamF42 committed:
[CST-18016] Added OpenAlexPublicationLoader as suggestionProvider

AdamF42 committed:
[CST-18016] Added generic script runner

AdamF42 committed:
[CST-18016] Updated tests to take into account newly created external sources

AdamF42 committed:
[CST-18016] Improved OpenAlex Publication mapping

AdamF42 committed:
[CST-18016] Improved doc

AdamF42 committed:
[CST-18016] Improved OpenAlex Publication mapping

AdamF42 committed:
[CST-18016] Improved OpenAlex Publication mapping adding dc.identifier.issn and oaire.citation.title

AdamF42 committed:
[CST-18016] Improved OpenAlex Person import mapping

AdamF42 committed:
[CST-18016] Fixed OpenAlexPublicationLoaderTest

AdamF42 committed:
[CST-18016] Improved OpenAlex Person name mapping

AdamF42 committed:
[CST-18016] Updated config external-providers.cfg to include OpenAlex sources

AdamF42 committed:
[CST-18016] Added OpenAlex Journal mapping

AdamF42 committed:
[CST-18016] Added other openalex urls

AdamF42 committed:
[CST-18016] Added Institution, Publisher and Funder mapping

AdamF42 committed:
[CST-18016] Updated OpenAlexAuthorNameContributor doc

AdamF42 committed:
[CST-18016] Fixed tests

AdamF42 committed:
[CST-18016] Fix on OpenAlexPublicationLoader key name

AdamF42 committed:
[CST-18016] Fix on CombinedMetadatumContributor to handle different list size

tdonohue commented:
@amgciadev : Has your team already reviewed/tested this work (as I see you helped coordinate the work)?  From my perspective, it'd be helpful if someone from your team could provide public feedback (either a code review or testing results or similar), as it can be beneficial in helping us to get this merged more rapidly.  Thanks!

amgciadev commented:
Hi @tdonohue, yes that is correct, Cambridge will be doing testing for this PR as part of this project. We are currently preparing the test instance.

pilasou commented:
Hi @AdamF42 , hi @tdonohue, I have tried to test the Live Import part of the PR. I made a search (OpenAlex Title field) and reach the preview window, but when I clic on Start submission button, I have an infinite load screen: 
![image](https://github.com/user-attachments/assets/e8955a56-2c89-456b-a452-fbe6efa9e6e1)

There are errors in the Browser DevTools console: 
![image](https://github.com/user-attachments/assets/728a38fe-f556-4692-98b4-53cfa739120d)

I am testing using DSpace Docker environnement on Windows (https://wiki.lyrasis.org/display/DSPACE/Testing+DSpace+Github+Pull+Requests)

AdamF42 committed:
[CST-18016] Fix on incorrect openalex publication id mapping

AdamF42 commented:
@pilasou thanks for testing the PR!

It appears that the issue was caused by an incorrect mapping of an OpenAlex ID for the Publication field. I have now corrected the mapping, so it should work as expected.

If the error persists, could you kindly attach a screenshot of the Network tab from your browser’s DevTools? Additionally, having the relevant logs would be extremely helpful for further investigation. Thank you in advance!

Also, if you haven’t already, I recommend importing dspace-types to ensure all necessary metadata fields are available. You can do so by running the following command inside your DSpace container:

```
docker container exec -it dspace bash -c "JAVA_TOOL_OPTIONS= && ./bin/dspace registry-loader -metadata config/registries/dspace-types.xml"
```

Let me know if you still encounter any issues!

tdonohue commented on dspace-api/src/main/java/org/dspace/app/suggestion/loader/PublicationLoader.java:
```diff
@@ -136,14 +138,16 @@ private Suggestion translateImportRecordToSuggestion(Item item, ExternalDataObje
-                + openAireId);
+                                            + "/api/integration/externalsources/" +
+                                            primaryProvider.getSourceIdentifier() + "/entryValues/"
+                                            + openAireId);
```
I like that you've made this class more generic, but I find it a bit odd that it still has a lot of references to OpenAIRE within it (even though the class is now a generic PublicationLoader).  This is an example, where this variable has the name `openAireId` even though it doesn't have anything to do with OpenAIRE...it's just an identifier and could be renamed `recordId`.

While I won't require it..it might be nice to clean up the other references to OpenAIRE in this file, as that will make it less confusing to developers looking at it in the future.

AdamF42 replied:
I’ve removed all references to OpenAIRE from PublicationLoader, including renaming variables like openAireId to more neutral terms such as recordId.

Additionally, I moved EvidenceScorer, AuthorNamesScorer, and DateScorer to a dedicated package, as they are generic and can be reused independently of OpenAIRE.

tdonohue commented on dspace-api/src/main/java/org/dspace/app/suggestion/runnable/PublicationLoaderRunnableCli.java:
```diff
@@ -0,0 +1,56 @@
+        // in case of CLI we show the help prompt
+        if (commandLine.hasOption('h')) {
+            HelpFormatter formatter = new HelpFormatter();
+            formatter.printHelp("Import Researcher Suggestions", getScriptConfiguration().getOptions());
```
This description ("Import Researcher Suggestion") is different from the one in the Processes User Interface which says: "Imports suggestions from external providers for publication claim"

I prefer the one displayed in the Processes UI, as it's more descriptive.

tdonohue commented on dspace-api/src/main/java/org/dspace/app/suggestion/script/PublicationLoaderCliScriptConfiguration.java:
```diff
@@ -0,0 +1,38 @@
+    public Options getOptions() {
+        Options options = super.getOptions();
+        options.addOption("h", "help", false, "help");
+        options.getOption("h").setType(boolean.class);
```
It's possible to run the `--help` command from the Processes UI, but it *fails* unless you also add the `--loader` command.  This behavior seems odd to me.

![processes-import](https://github.com/user-attachments/assets/77f9bd72-70ce-4f9d-a918-a0d3c692d46d)

I'd expect you to be able to run `--help` with no additional parameters.

AdamF42 replied:
Thank you for your feedback!

You’re right that the current behavior seems odd. There are two aspects to consider:

1. **Frontend Handling**: The check for required parameters is currently performed only on the frontend. It might be better to fix this at the frontend level by making an exception for the --help flag, as it is a special case. This issue also affects other processes requiring mandatory parameters, such as **bulk-access-control**.

2. **Backend Execution**: Even if the help command runs, the _ScriptRestRepository#runDSpaceScript_ method does not take into account the result of the runnable initialization (dSpaceRunnable.initialize(...)). This means that the script is scheduled regardless of whether initialization was successful.
To address this, we might consider modifying the implementation to check the return value of the initialization step, similar to the following approach:
```
DSpaceRunnable.StepResult result =
    dSpaceRunnable.initialize(args.toArray(new String[0]), restDSpaceRunnableHandler,
                              context.getCurrentUser());
if (result.equals(DSpaceRunnable.StepResult.Continue)) {
    if (files != null && !files.isEmpty()) {
        checkFileNames(dSpaceRunnable, files);
        processFiles(context, restDSpaceRunnableHandler, files);
    }
    restDSpaceRunnableHandler.schedule(dSpaceRunnable);
}
```

tdonohue commented on dspace/config/spring/api/scripts.xml:
```diff
@@ -102,4 +102,10 @@
        <property name="dspaceRunnableClass" value="org.dspace.app.suggestion.openaire.PublicationLoaderRunnableCli"/>
     </bean>
 
+    <bean id="import-loader-suggestions"
```
Is this new script meant to **replace** the existing `import-openaire-suggestions` script?  To me, it seems like the `import-openaire-suggestions` script is no longer necessary.  Should we remove it and point at using `import-loader-suggestions` instead?

tdonohue commented on dspace-api/src/main/java/org/dspace/app/suggestion/script/PublicationLoaderScriptConfiguration.java:
```diff
@@ -0,0 +1,105 @@
+            options.addOption("f", "solrfilter", true, "The additional SORL filter to better refine results");
+            options.getOption("f").setType(String.class);
+
+            options.addOption("il", "limit", true, "the max number of Researcher profiles. If no limit is provided, "
```
I'm not a fan of a two character option.  `-il`, especially since this could be easily confused with your `-l` option above.

Could we consider renaming this to be `-m` / `--max`?  Then the description just needs minor updates to say: `The maximum number of researcher profiles to process. If no maximum is provided, then the configured default will be used.`

AdamF42 replied:
I renamed the "il/limit" option to "m/max".

tdonohue commented on dspace-api/src/main/java/org/dspace/app/suggestion/runnable/PublicationLoaderRunnable.java:
```diff
@@ -0,0 +1,200 @@
+     * @return The default item limit, as defined in the DSpace configuration.
+     */
+    private Integer getDefaultLimit() {
+        return configurationService.getIntProperty("publication-loader.limit", 1000);
```
This configuration doesn't appear in any `*.cfg` file.  It needs to be added to a config with some basic documentation about it's purpose (and default value)

It seems like it might be best to go in the `suggestion.cfg`?  If that's the case, maybe we should rename this to be:
```
suggestion.publication-loader.max = 1000
```

(I don't mind the name `limit`, but if we rename the flag to be `--max`, then it might make sense to change the name here as well)

AdamF42 replied:
Added suggestion.publication-loader.max to suggestion.cfg.

tdonohue commented on dspace-api/src/main/java/org/dspace/discovery/DiscoverResultIterator.java:
```diff
@@ -0,0 +1,144 @@
+        this(context, null, discoverQuery, true, -1);
+    }
+
+    public DiscoverResultIterator(Context context, DiscoverQuery discoverQuery, boolean uncacheEntities) {
```
Please add JavaDocs to all the public constructors & methods in this class

tdonohue commented on dspace-api/src/main/java/org/dspace/discovery/DiscoverResultIterator.java:
```diff
@@ -0,0 +1,144 @@
+    }
+
+    @Override
+    public boolean hasNext() {
```
We should also add JavaDocs to these override methods to describe the overridden behavior

tdonohue commented on dspace-api/src/main/java/org/dspace/discovery/DiscoverResultIterator.java:
```diff
@@ -0,0 +1,144 @@
+ * @author Luca Giamminonni (luca.giamminonni at 4science.it)
+ */
+@SuppressWarnings({"rawtypes", "unchecked"})
+public class DiscoverResultIterator<T extends ReloadableEntity, PK extends Serializable> implements Iterator<T> {
```
It may be nice to add some basic tests (either Unit or Integration tests) for both `DiscoverResultIterator` and `DiscoverResultItemIterator`, as I'm seeing a **lot** of warnings from our Codecov scanner in these classes saying this code is untested.

I like the design of each of these iterators, but adding automated tests would help ensure they are behaving properly.

tdonohue commented on dspace-api/src/main/java/org/dspace/importer/external/openalex/metadatamapping/OpenAlexDateMetadataProcessor.java:
```diff
@@ -0,0 +1,96 @@
+
+        try {
+            String dateStr = node.asText();
+            LocalDate date = LocalDate.parse(dateStr, DateTimeFormatter.ISO_DATE);
```
Yes! Thanks for using `java.time.*` here :)  I'm working to get us off of `java.util.Date` in #10432 , so it's great to see you are also using `java.time.*` in this PR.

tdonohue commented on dspace-server-webapp/src/main/java/org/dspace/app/rest/converter/RootConverter.java:
```diff
@@ -31,7 +32,8 @@ public RootRest convert(HttpServletRequest request) {
         String dspaceSSRUrl = configurationService.getProperty("dspace.server.ssr.url", dspaceUrl);
-        if (!dspaceUrl.equals(dspaceSSRUrl) && requestUrl.startsWith(dspaceSSRUrl)) {
+        if (!dspaceUrl.equals(dspaceSSRUrl) && StringUtils.isNotBlank(dspaceSSRUrl)
+            && requestUrl.startsWith(dspaceSSRUrl)) {
```
This change doesn't seem related to OpenAlex.  I'm not against it, as it seems "logical" to verify that the request starts with the configured SSR URL.  But, I wanted to ask if this was a bug you ran into related to OpenAlex or something more general?  

The reason I wanted to ask is that **if this is a general bug**, then we may need to ensure this small fix gets **backported to 8.x and 7.6.x**.  So, it may need to be pulled out into a separate PR specific to 8.x/7.x, because we won't be backporting OpenAlex integration

tdonohue commented on dspace-server-webapp/src/test/resources/org/dspace/app/rest/openalex-institution-multiple.json:
```diff
@@ -0,0 +1,5575 @@
+{
```
This is a rather large test file (over 5,000 lines).  Would it be possible to trim this down to *only* the content we need for testing?  It just seems like a very large file for us to store just for testing.

tdonohue commented on dspace-server-webapp/src/test/resources/org/dspace/app/rest/openalex-institution-single.json:
```diff
@@ -0,0 +1,4426 @@
+{
+  "meta": {
```
Same here, this is a very large test file (>4,000 lines).  For all of these test JSON files that are >1,000 lines, I'd like to know if we could trim them down.  This PR is over 27,000 lines of code, but it looks like >21,000 lines are just these very large test files.  So, the PR itself would be much smaller if we could simply trim these files down in size.

tdonohue reviewed (changes requested):
@AdamF42 : Thank you for this work to integrate OpenAlex!  I gave this an initial code review today, and overall it looks great!   I do, however, have some inline code suggestions / places where I feel this PR could be cleaned up.  None of them seem major to me. But, if you find any are difficult or disagree with my suggestions, please do let me know.  

The one suggestion I feel most strongly about is to *try to please make the tests content files smaller*.  I'm noticing that >21,000 lines of this 27,000 line PR is just test files. That means our test files are too big :) 

I also gave this some light testing by importing a few Items from OpenAlex.  It seems to be working well & I haven't noticed any obvious bugs.  I haven't yet tested the Publication Claim, but I plan to do so soon.  I can report back on whether I find any issues there.

amgciadev commented:
Hi @AdamF42 I am also starting testing of this PR now. I have deployed the pr locally (I am not using docker) but when I try to import a publication via the import sources I don't see the openalex options, I only see openaire options - see screenshot. Is there any other config I am missing?
![Uploading Screenshot 2025-03-11 at 14.29.41.png…]()

tdonohue commented:
@amgciadev : That list of import options *should be scrollable*.  At least it is for me & it works that way on our https://sandbox.dspace.org site.  So, I don't think this is a bug in DSpace itself, but maybe you have custom code causing this issue?

tdonohue commented:
@amgciadev : I'd also recommend trying https://sandbox.dspace.org to make sure it works there for you.  This PR doesn't seem to change the behavior of that scrollable box...and sandbox.dspace.org is running latest `main`.

I can verify on my end I'm seeing it scrollable on latest `main` (and in this PR) on Chrome.

amgciadev commented:
All sorted. No bug. I have tested the import with a few publications and it also works well for me.

AdamF42 committed:
[CST-18016] Removed references from openAire in PublicationLoader

AdamF42 committed:
[CST-18016] Updated PublicationLoaderRunnableCli description

amgciadev commented:
I have now testing the publications suggestion loader with a single researcher, and after having to play a bit with matching names, I have successfully obtained matching suggestions.

![Screenshot 2025-03-12 at 07 56 22](https://github.com/user-attachments/assets/ef51b6e3-a71f-4112-857c-ec1d8e79f313)


It currently is quite strict: my name has accents, e.g. Martínez García and unless I was putting those under dc.title for the researcher profile it wouldn't come up with any matches, even though there is a match on OpenAlex author ID.

AdamF42 committed:
[CST-18016] Reduced the size of the test content files

AdamF42 committed:
[CST-18016] Fixed OpenAlex Search by Author ID

AdamF42 commented on dspace/config/spring/api/openalex-integration.xml:
```diff
@@ -0,0 +1,851 @@
+        <entry key-ref="oaire.citation.title" value-ref="openalexConditionalPrimaryLocationTitleContrib"/>
+        <entry key-ref="dc.identifier.issn" value-ref="openalexConditionalPrimaryLocationISSNContrib"/>
+        <entry key-ref="openalex.type" value-ref="openalexTypeContrib"/>
+    </util:map>
```
@amgciadev  Thanks for the suggestion!

The `best_oa_location/pdf_url` is already mapped under dc.identifier.uri. In fact, to be more precise, the following fields are all mapped to dc.identifier.uri:

    /best_oa_location/landing_page_url
    /best_oa_location/pdf_url
    /locations/landing_page_url
    /locations/pdf_url
    /primary_location/landing_page_url
    /primary_location/pdf_url
    /open_access/oa_url

That said, it is definitely possible to create a new metadata field in the registry specifically for full-text links. Would it perhaps be better to use [oa_url](https://docs.openalex.org/api-entities/works/work-object#oa_url) for this purpose? Let me know your thoughts!

AdamF42 replied:
@amgciadev, apologies for missing your comment earlier!

In the `openalex-integration.xml`, the `openalexUriIdentifierscContrib` maps the JSON path to `dc.identifier.uri`. Since this is a `MultipleMetadataContributor`, it aggregates multiple contributors into a list under the same metadata field.

The `contribList` defines how each element in the result list is mapped to `dc.identifier.uri`.

In `submission-forms.xml`, `dc.identifier` is a repeatable field, so it will be displayed like this:

![image](https://github.com/user-attachments/assets/9fb60648-3298-4560-844f-d817a31bbf8a)

Would you like to map it into a dedicated metadata field instead?

AdamF42 replied:
After an internal discussion, we determined that `dc.identifier.uri` is not the appropriate field for these mappings, as it should be used exclusively for the URI of the item within DSpace. 
Since DSpace does not have a dedicated metadata field that accurately represents these external identifiers, we decided to retain only the Best OA Location (PDF and Landing Page) mappings as follows:

- `/best_oa_location/landing_page_url` → `dc.identifier.other`
- `/best_oa_location/pdf_url` → `dc.identifier.other`

Here’s how it looks now:
![image](https://github.com/user-attachments/assets/a5689f73-35e5-4e5d-a676-4396f95f01d2)

The last two identifiers correspond to the landing page URL and PDF URL for the OpenAlex work [W2046182783](https://api.openalex.org/works/W2046182783).

Additionally, I updated MultipleMetadataContributor to exclude duplicates and added unit tests to ensure the changes are correctly applied.

AdamF42 commented:
> I have now tested the publications suggestion loader with a single researcher, and after having to play a bit with matching names, I have successfully obtained matching suggestions.
> 
> ![Screenshot 2025-03-12 at 07 56 22](https://private-user-images.githubusercontent.com/26137686/421764118-ef51b6e3-a71f-4112-857c-ec1d8e79f313.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDE4MjUyNjUsIm5iZiI6MTc0MTgyNDk2NSwicGF0aCI6Ii8yNjEzNzY4Ni80MjE3NjQxMTgtZWY1MWI2ZTMtYTcxZi00MTEyLTg1N2MtZWMxZDhlNzlmMzEzLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTAzMTMlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwMzEzVDAwMTYwNVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWM4MTNkNzZhYzQ5NTdmNjFjYzkwN2NmZTMzYzlhN2M0YWE4MmMxNjZkMGQ3MDY0MTdiMzkxMGZmOTBmNDEwZTcmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.UG8oXqOmbpPjbk042KPA9zRAWimly7KSpTuCQ5BE4Rg)
> 
> It currently is quite strict: my name has accents, e.g. Martínez García and unless I was putting those under dc.title for the researcher profile it wouldn't come up with any matches, even though there is a match on OpenAlex author ID.

Thanks for testing the publication suggestion loader!

I’d like to better understand the issue you encountered. Could you provide more details about the steps to reproduce it? Specifically, how were you setting up the researcher profile before running the loader?

Also, just to clarify, the `import-loader-suggestion` process using OpenAlex can only work if the researcher has been imported from OpenAlex or has a metadata value set for `dc.identifier.openalex`. This limitation exists because the OpenAlex API does not allow searching for publications directly by author name. 
Let me know if this was the case in your test!

AdamF42 committed:
[CST-18016] Updated openalex lastimport metadata.

AdamF42 commented on dspace/config/registries/dspace-types.xml:
```diff
@@ -93,4 +93,11 @@
+        <schema>dspace</schema>
+        <element>lastimport</element>
+        <qualifier>loader-openalex</qualifier>
+        <scope_note>Stores the timestamp related to the last import from OpenAlex</scope_note>
```
I definitely agree!

amgciadev commented:
Hi I am testing the import publication search by author ID. I wanted to double-check that the Open Alex ID is the one to use here. If so, I am not retrieving any results. E.g. I am using the ID A5102019421, and the equivalent query in OpenAlex retrieves results as expected: https://api.openalex.org/works?filter=author.id:A5102019421

I have also tried with another ID: A5090128810, and the same behaviour: no results in DSpace but the associated openalex query retrieves 112 results as expected (https://api.openalex.org/works?filter=author.id:A5090128810)

AdamF42 committed:
[CST-18016] Fixed AuthorNamesScorer to replace diacritical marks during normalization

AdamF42 commented:
> > > I have now tested the publications suggestion loader with a single researcher, and after having to play a bit with matching names, I have successfully obtained matching suggestions.
> > > ![Screenshot 2025-03-12 at 07 56 22](https://private-user-images.githubusercontent.com/26137686/421764118-ef51b6e3-a71f-4112-857c-ec1d8e79f313.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDE4MjUyNjUsIm5iZiI6MTc0MTgyNDk2NSwicGF0aCI6Ii8yNjEzNzY4Ni80MjE3NjQxMTgtZWY1MWI2ZTMtYTcxZi00MTEyLTg1N2MtZWMxZDhlNzlmMzEzLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTAzMTMlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwMzEzVDAwMTYwNVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWM4MTNkNzZhYzQ5NTdmNjFjYzkwN2NmZTMzYzlhN2M0YWE4MmMxNjZkMGQ3MDY0MTdiMzkxMGZmOTBmNDEwZTcmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.UG8oXqOmbpPjbk042KPA9zRAWimly7KSpTuCQ5BE4Rg)
> > > It currently is quite strict: my name has accents, e.g. Martínez García and unless I was putting those under dc.title for the researcher profile it wouldn't come up with any matches, even though there is a match on OpenAlex author ID.
> > 
> > 
> > Thanks for testing the publication suggestion loader!
> > I’d like to better understand the issue you encountered. Could you provide more details about the steps to reproduce it? Specifically, how were you setting up the researcher profile before running the loader?
> > Also, just to clarify, the `import-loader-suggestion` process using OpenAlex can only work if the researcher has been imported from OpenAlex or has a metadata value set for `dc.identifier.openalex`. This limitation exists because the OpenAlex API does not allow searching for publications directly by author name. Let me know if this was the case in your test!
> 
> Hi @AdamF42 , this is how I tested:
> 
>     1. Created a profile based on my existing user
> 
>     2. I added manually the dc.identifier.openalex ID to my profile (it took me a bit to figure it out but it makes sense!)
> 
>     3. Then run the import-loader-suggestion command
> 
> 
> Originally the name in my profile (dc.title metadata) didn't fully match the information in OpenAlex: I had "Agustina Martinez Garcia", and in the returned openalex publications my name was coming up as "Agustina Mart**í**nez Garc**í**a" (note the accents). This small difference was causing the evidence checks in the code to not match and therefore the suggestions were not created. I can confirm that a small variant of "Agustina Martínez-García" did also produce a match, but the lack of accents in the name was causing it for it to fail completely. It seems a bit too strict to me, but I don't know enough about the current matching algorithm.
> 
> I hope this clarifies and is helpful. I don't think is a bug per se but perhaps too strict criteria when matching by name, and I can see this causing issues for many authors.
> 
> Also thank you for the pointer of importing the profile first from OpenAlex, I'll try that out with other researchers and report back.

Thanks for the update! It looks like the issue you encountered was related to how the AuthorNamesScorer handled author names with diacritic characters (e.g., "é", "á", "ù", "í"). 

I’ve addressed this by adding a fix that ensures diacritical marks are properly removed during string normalization.

@tdonohue , if this fix looks good to you, do you think we should backport it to older versions?

AdamF42 committed:
[CST-18016] Updated dspace-types to include dspace.openaire.lastimport

AdamF42 committed:
[CST-18016] Removed import-openaire-suggestions script as it can be used with import-loader-suggestions script

AdamF42 committed:
[CST-18016] Renamed import-loader-suggestions limit parameter to max

AdamF42 committed:
[CST-18016] Added suggestion.publication-loader.limit to suggestion.cfg

AdamF42 committed:
[CST-18016] Improved DiscoverResultItemIterator and DiscoverResultIterator java doc

AdamF42 committed:
[CST-18016] Better max option description.

AdamF42 committed:
[CST-18016] Renamed suggestion.publication-loader.limit property to suggestion.publication-loader.max.

AdamF42 committed:
[CST-18016] Added missing doc for DiscoverResultIterator#getTotalSearchResults

AdamF42 committed:
[CST-18016] Added unit tests for DiscoverResultIteratorTest

AdamF42 committed:
[CST-18016] Added unit tests for DiscoverResultItemIteratorTest

tdonohue commented on dspace-api/src/main/java/org/dspace/importer/external/metadatamapping/contributor/ConditionEqualJsonPathMetadataContributor.java:
```diff
@@ -0,0 +1,94 @@
+    public Collection<MetadatumDTO> contributeMetadata(String json) {
+        Collection<String> leftOperands = leftOperandProcessor.processMetadata(json);
+        if (leftOperands == null || leftOperands.size() != 1) {
+            log.warn("No left operands found in json path: " + json);
```
We may not want to print out the entire JSON contents here.  This fills up the logs quickly when it is triggered. 

We might also want to create some automated tests for this case, as it appears to (possibly) not be working based on my manual testing.

amgciadev replied:
@tdonohue I can confirm that for all the author profiles I've tested with I always also get those warnings but suggestions are being loaded as expected. I have replicated your steps, with your researcher profile and OpenAlex ID, and I can confirm that I get 15 suggestions loaded for you, so a bit
![Screenshot 2025-03-18 at 07 55 12](https://github.com/user-attachments/assets/6ab3206b-b770-4450-8d08-7e719538ec7c)
![Screenshot 2025-03-18 at 07 55 05](https://github.com/user-attachments/assets/730a3ba7-1417-4e03-94c8-f3e8fa03befc)
 odd.

tdonohue replied:
@amgciadev : thanks for sharing your testing.  One thing that immediately jumps out at me is that I was setting `dc.identifier.openalex` to the full URL of `https://openalex.org/authors/a5103254513`, but it looks like you've set it to just the identifier of `a5103254513`

So, It's possible that I was misunderstanding **how to test this**.  I'll see if I can find time to try it again today. 

@AdamF42 : I still feel we should **not** log the entire JSON file on the line above, as it fills up the logs very quickly with (seemingly) useless information.  I also would still like to see more automated tests for `ConditionEqualJsonPathMetadataContributor.java`, if it's possible to create more.  It's unclear to me why it logs this warning multiple times for every external "suggestion"

AdamF42 replied:
@tdonohue thanks for your feedback!  

I agree that logging the entire JSON file was excessive. I've updated the log level to **DEBUG**, as it can still be helpful for troubleshooting but won't clutter the logs unnecessarily.  

The reason why the warning message appears six times per suggestion is due to how the mapping is structured. Specifically, it attempts to map `oaire.version`, `dc.identifier.issn`, and `oaire.citation.title` as follows:  

  - `openalexPublicationVersionContrib` maps **`oaire.version`** with:  
    - `openalexConditionalPublishedVersionContrib`: Sets `oaire.version` to `http://purl.org/coar/version/c_970fb48d4fbd8a85` if `/best_oa_location/version = publishedVersion`  
    - `openalexConditionalAcceptedVersionContrib`: Sets `oaire.version` to `http://purl.org/coar/version/c_fa2ee174bc00049f` if `/best_oa_location/version = acceptedVersion`  
    - `openalexConditionalSubmittedVersionContrib`: Sets `oaire.version` to `http://purl.org/coar/version/c_b1a7d7d4d402bcce` if `/best_oa_location/version = submittedVersion`  

  - `openalexConditionalPrimaryLocationTitleContrib` maps **`oaire.citation.title`** with:  
    - `openalexConditionalPrimaryLocationTitleJourContrib`: Sets `oaire.citation.title` to `/primary_location/source/display_name` if `/primary_location/source/type = journal`  
    - `openalexConditionalPrimaryLocationTitleConfContrib`: Sets `oaire.citation.title` to `/primary_location/source/display_name` if `/primary_location/source/type = conference`  

  - `openalexConditionalPrimaryLocationISSNContrib` maps **`dc.identifier.issn`** with:  
    - Sets `dc.identifier.issn` to `/primary_location/source/issn` if `/primary_location/source/type = journal`  

In our specific case, the API call to [`https://api.openalex.org/works?filter=authorships.author.id:A5103254513`](https://api.openalex.org/works?filter=authorships.author.id:A5103254513) returns **15 elements**. However, except for the first (`W1549954474`) and the last (`W44159624`), both `/best_oa_location/version` and `/primary_location/source/type` are **always null** result in the warn message you posted.

Regarding test coverage, these mappings are **partially covered** in the **`Openalex*ExternalSourcesIT`** integration tests. Additionally, I have added a **dedicated unit test** (`ConditionEqualJsonPathMetadataContributorTest`) to demonstrate the basic usage of `ConditionEqualJsonPathMetadataContributor`.  

Let me know if you have any other suggestions or concerns!

tdonohue replied:
I've retested today and found that suggestions are now importing properly from OpenAlex.  The issue was my value of `dc.identifier.openalex`.

tdonohue reviewed (changes requested):
@AdamF42 : Thanks!  Overall the code looks great.  I retested things today, and the Import from OpenAlex in MyDSpace is working well.

However, I'm having difficulty getting the `import-loader-suggestions` script to work properly to create suggestions for Publication Claim from OpenAlex.

Here's what I've done:
1. Locally, I have an account for myself with a Researcher Profile (attached Person Entity)
2. I edited that Researcher Profile and added a `dc.identifier.openalex` field with a value of https://openalex.org/authors/a5103254513  (This is an identifier for some of my old works in OpenAlex)
3. Then, as an Admin, I ran the `import-loader-suggestions` script from the Admin UI with `--loader openalex` and `--single-researcher [uuid-of-my-person-entity]`.
4. The script succeeds, but nothing appears under Publication Claim (Notifications -> Publication Claim -> OpenAlex)
5. If I check the backend logs, I see a very large number of warnings that look like this:
```
2025-03-17 20:02:40,072 WARN  unknown unknown org.dspace.importer.external.metadatamapping.contributor.ConditionEqualJsonPathMetadataContributor @ No left operands found in json path: [Large JSON output]
2025-03-17 20:02:40,072 WARN  unknown unknown org.dspace.importer.external.metadatamapping.contributor.ConditionEqualJsonPathMetadataContributor @ No left operands found in json path: [Large JSON output
....
025-03-17 20:02:40,101 WARN  unknown unknown org.dspace.importer.external.metadatamapping.contributor.InvertedIndexProcessor @ The specified path /abstract_inverted_index is not a JSON array
```
Each of the "No left operands found in json path" appears to be repeated six times for every work that I have listed at https://openalex.org/works?page=1&filter=authorships.author.id%3Aa5103254513  

For each of the warnings a **large amount of JSON is also printed out in the logs** six times in a row.

Based on what I see in the logs, it looks like it's found my OpenAlex record and is attempting to import my works, but every work fails to be imported?

Or am I missing something here with how the `import-loader-suggestions` script is supposed to work?

amgciadev reviewed (changes requested):
@AdamF42 thanks for the PR, the code looks good, and the fixes to the diacritics in names are now working well. I have now done some more testing and have uncovered a bug that affects the import of records both when running the import-suggestions-loader and also when using the LiveImport and searching by author ID: suggestions are presented as expected but when importing records no information is found and this causes for items to be created in the workspace with no metadata filled out.

amgciadev commented on dspace-api/src/main/java/org/dspace/importer/external/openalex/service/OpenAlexImportMetadataSourceServiceImpl.java:
```diff
@@ -0,0 +1,366 @@
+    private List<ImportRecord> searchById(String id) {
+        List<ImportRecord> results = new ArrayList<>();
+        try {
+            String resp = liveImportClient.executeHttpGetRequest(timeout, this.url + "/" + id, new HashMap<>());
```
@AdamF42 the bug I've mentioned is in this method (searchById): when searching by author ID the query URL is in the form https://api.openalex.org/works?filter=authorships.author.id:, and this method expects it to be without filters to directly append the open alex work ID to retrieve a single work. The resulting URL ends up being: https://api.openalex.org/works?filter=authorships.author.id/W3170298948:, which returns 0 results, and thus the metadata import is empty

AdamF42 replied:
Thanks for pointing that out! I added an integration test to replicate the issue and applied a fix to ensure the correct query URL is used. 
Let me know if you encounter any further issues!

amgciadev reviewed (changes requested):
In addition to the above mentioned bug, see a couple of suggestions plus a minor bug in the angular side:

- As per your comment here: https://github.com/DSpace/DSpace/pull/10433#discussion_r1992499373 would it be possible to add the Best OA field to the map of imported metadata so that is both presented to the user in the live import preview, as well as added to the record? At present you mentioned that it is mapped to dc.identifirt.uri but it is not presented to the user as far as I can tell. It is useful to have it in the preview as the user can then download the file at import time, when creating the workspace item.
- Small bug when using the import-suggestion-loader with the max researcher profiles parameter (-m). The UI does not present an input field and therefore you cannot enter the option value and the script fails. I am reporting in the angular PR.
- **[Optional but very useful]**. Would you consider adding the ability to search by DOI in addition to the current "search by title" and "search by author id" options? Searching by DOI is a very common search option that researchers use all the time, plus it returns accurate, individual results. The required code changes to add this would be minor, and mostly done via configuration. See some snippets below.

In spring-dspace-addon-import-services.xml:

```
<bean id="openalexImportPublicationByDOIService"
          class="org.dspace.importer.external.openalex.service.OpenAlexImportMetadataSourceServiceImpl">
        <property name="metadataFieldMapping" ref="openalexPublicationMetadataFieldMapping"/>
        <property name="url" value="${openalex.url.works}?filter=doi:"/>
    </bean>
```

In external-services.xml:

```
<bean id="openalexPublicationByDOILiveImportDataProvider"
          class="org.dspace.external.provider.impl.LiveImportDataProvider">
        <property name="metadataSource" ref="openalexImportPublicationByDOIService"/>
        <property name="sourceIdentifier" value="openalexPublicationByDOI"/>
        <property name="recordIdMetadata" value="dc.identifier.openalex"/>
        <property name="supportedEntityTypes">
            <list>
                <value>Publication</value>
                <value>none</value>
            </list>
        </property>
    </bean>
```

In OpenAlexImportMetadataSourceServiceImpl.java, modify **search and count** methods to update the query url similar to what is already done for the author id search:

```
private List<ImportRecord> search(String query, Integer page, Integer pageSize) {
...
if (queryUrl.contains("?filter=authorships.author.id:")) {
                queryUrl = queryUrl.replace("?filter=authorships.author.id:", "");
                uriParams.put("filter", "authorships.author.id:" + query);
            } else if (queryUrl.contains("?filter=doi:")) {
                queryUrl = queryUrl.replace("?filter=doi:", "");
                uriParams.put("filter", "doi:" + query);
            }
```

AdamF42 committed:
[CST-18016] Dummy naming fix

AdamF42 committed:
[CST-18016] Added test to reproduce the bug with OpenAlexImportMetadataSourceServiceImpl#searchById

AdamF42 committed:
[CST-18016] Fixed OpenAlexImportMetadataSourceServiceImpl#searchById when called from search by author id

AdamF42 committed:
[CST-18016] Added basic unit test for ConditionEqualJsonPathMetadataContributor

AdamF42 committed:
[CST-18016] Updated ConditionEqualJsonPathMetadataContributor log level from warn to debug

AdamF42 committed:
[CST-18016] Added openalexImportPublicationByDOIService

AdamF42 committed:
[CST-18016] Added tests for OpenAlexPublicationByDOIExternalSourcesIT

AdamF42 committed:
[CST-18016] Updated OpenAlexImportMetadataSourceServiceImpl to search publications by doi

AdamF42 committed:
[CST-18016] removed useless test

AdamF42 commented:
> In addition to the above mentioned bug, see a couple of suggestions plus a minor bug in the angular side:
> 
>     * As per your comment here: [OpenAlex Integration #10433 (comment)](https://github.com/DSpace/DSpace/pull/10433#discussion_r1992499373) would it be possible to add the Best OA field to the map of imported metadata so that is both presented to the user in the live import preview, as well as added to the record? At present you mentioned that it is mapped to dc.identifirt.uri but it is not presented to the user as far as I can tell. It is useful to have it in the preview as the user can then download the file at import time, when creating the workspace item.
> 
>     * Small bug when using the import-suggestion-loader with the max researcher profiles parameter (-m). The UI does not present an input field and therefore you cannot enter the option value and the script fails. I am reporting in the angular PR.
> 
>     * **[Optional but very useful]**. Would you consider adding the ability to search by DOI in addition to the current "search by title" and "search by author id" options? Searching by DOI is a very common search option that researchers use all the time, plus it returns accurate, individual results. The required code changes to add this would be minor, and mostly done via configuration. See some snippets below.
> 
> 
> In spring-dspace-addon-import-services.xml:
> 
> ```
> <bean id="openalexImportPublicationByDOIService"
>           class="org.dspace.importer.external.openalex.service.OpenAlexImportMetadataSourceServiceImpl">
>         <property name="metadataFieldMapping" ref="openalexPublicationMetadataFieldMapping"/>
>         <property name="url" value="${openalex.url.works}?filter=doi:"/>
>     </bean>
> ```
> 
> In external-services.xml:
> 
> ```
> <bean id="openalexPublicationByDOILiveImportDataProvider"
>           class="org.dspace.external.provider.impl.LiveImportDataProvider">
>         <property name="metadataSource" ref="openalexImportPublicationByDOIService"/>
>         <property name="sourceIdentifier" value="openalexPublicationByDOI"/>
>         <property name="recordIdMetadata" value="dc.identifier.openalex"/>
>         <property name="supportedEntityTypes">
>             <list>
>                 <value>Publication</value>
>                 <value>none</value>
>             </list>
>         </property>
>     </bean>
> ```
> 
> In OpenAlexImportMetadataSourceServiceImpl.java, modify **search and count** methods to update the query url similar to what is already done for the author id search:
> 
> ```
> private List<ImportRecord> search(String query, Integer page, Integer pageSize) {
> ...
> if (queryUrl.contains("?filter=authorships.author.id:")) {
>                 queryUrl = queryUrl.replace("?filter=authorships.author.id:", "");
>                 uriParams.put("filter", "authorships.author.id:" + query);
>             } else if (queryUrl.contains("?filter=doi:")) {
>                 queryUrl = queryUrl.replace("?filter=doi:", "");
>                 uriParams.put("filter", "doi:" + query);
>             }
> ```

@amgciadev Thanks for your suggestions!
I addressed the first two points in the following discussions:
- [Metadata mapping](https://github.com/DSpace/DSpace/pull/10433#discussion_r2002650710)
- [Bug in import-suggestion-loader with -m parameter](https://github.com/DSpace/dspace-angular/pull/4017#issuecomment-2732563577)

Regarding the third point, I added integration tests and implemented the search by DOI. I used `DoiCheck` to clean up the DOI input, similar to how it is handled in `CrossRefImportMetadataSourceServiceImpl` and `ScopusImportMetadataSourceServiceImpl`.

This means it is now possible to search using a DOI in various formats, such as:
- `10.1016/s0021-9258(19)52451-6`
- `https://doi.org/10.1016/s0021-9258(19)52451-6`

amgciadev reviewed (approved):
@AdamF42, thanks so much for addressing all the comments and also implementing the search by DOI addition - which will be very useful for researchers. I have now retested again all the different search options, as well as the import-suggestions--loader script and can confirm that all is working as expected. I have also had a look at the code additions and they all look good to me.

AdamF42 committed:
[CST-18016] updated MultipleMetadataContributor to avoid duplicates

AdamF42 committed:
[CST-18016] Updated tests to take into account the new desired mapping

AdamF42 committed:
[CST-18016] Removed mapping on dc.identifier.uri and used dc.identifier.other

tdonohue reviewed (approved):
👍 Thanks @AdamF42 !  This now all looks good to me now.  I gave it another test today and was able to successfully import myself as a new `Person` entity from OpenAlex, then pull in Suggestions from OpenAlex for that `Person` entity.  I was also able to simply import a new publication from OpenAlex as well.

The recent code changes all look good too!

I'm flagging this as "Needs Documentation" as we will need to add some basic docs about OpenAlex to the 9.0 Docs at https://wiki.lyrasis.org/display/DSDOC9x/   Specifically, we likely need to add docs on OpenAlex at https://wiki.lyrasis.org/display/DSDOC9x/Publication+Claim and https://wiki.lyrasis.org/pages/viewpage.action?pageId=379125956

Finally, just a reminder that we need to backport *two* bug fixes that were applied in this PR, see https://github.com/DSpace/DSpace/pull/10433#discussion_r1988071503 and https://github.com/DSpace/DSpace/pull/10433#discussion_r1999311746  Please create a PR to backport these to 8.x & 7.x, and I'll gladly review it.

tdonohue commented:
@AdamF42 or @atarix83 :  Would someone at 4Science be able to add documentation for this feature in our 9.0 docs? https://wiki.lyrasis.org/display/DSDOC9x   Ideally, basic documentation would exist prior to our [9.0 Testathon](https://wiki.lyrasis.org/display/DSPACE/DSpace+Release+9.0+Testathon+Page) (which starts on Monday, April 7).  But, if it cannot exist by then, it should hopefully be created shortly thereafter.

AdamF42 commented:
> @AdamF42 or @atarix83 : Would someone at 4Science be able to add documentation for this feature in our 9.0 docs? https://wiki.lyrasis.org/display/DSDOC9x Ideally, basic documentation would exist prior to our [9.0 Testathon](https://wiki.lyrasis.org/display/DSPACE/DSpace+Release+9.0+Testathon+Page) (which starts on Monday, April 7). But, if it cannot exist by then, it should hopefully be created shortly thereafter.

I will provide the basic documentation by the end of the week.

tdonohue commented:
Documented now at https://wiki.lyrasis.org/display/DSDOC9x/Live+Import+from+external+sources#LiveImportfromexternalsources-OpenAlexIntegration and https://wiki.lyrasis.org/display/DSDOC9x/Publication+Claim
